### DISCUSSION TITLE
Element Activation gets stuck on "Handling Mesh Changes" after 76539 steps, DT, core count and mesh independent

### DISCUSSION THREAD

**Discussion by Edward-Eth** [idaholab/moose#22868] [Q&A General] [answered]
As the title says, I am testing out various components of MOOSE to find issues I've been experiencing in larger simulations. One thing I've been checking is the long-duration/high step stability of various functionalities. 

As part of this I've been running my 'simulation' physics-less, just with the element activation component active. This means in the model no physics is occurring but the mesh is being continually changed according to my input file.

My "goal" total sim time is around 1 million seconds. Initially I ran at a dt of 10, and the simulation failed at 765390 seconds, or 76539 steps. However, when I start at 760000 seconds, it runs to the end without issue. When I start at zero but with a time step of 5 instead of 10, I again fail at 76539 steps. This means there is no link between the issue and the state of my mesh/simulation, just the time step number itself...

This is very odd behaviour that I am at a loss to explain. 76539 isn't even a power of 2!

Any advice would be much appreciated!

**Comment by GiudGiud** [comment]
Hello

Is the memory consumption growing? (There s postprocessors and VPP) to measure that 
Can you share your input file?

Guillaume

**Reply by Edward-Eth** [reply]
I'm running now with memory postprocessors, will update. 

I can potentially share the input file, but in essence all I'm doing is:
- loading a mesh from a file
- setting up "timer" aux variables on each element of the mesh which count down (ie designated_Time - time)
- using ActivateElementsCoupled on the timer aux variables to move elements from block 0 to block 1 when the timer goes below 0
-  I have a few variables defined, but no active kernels, BCs etc so that the residual is always zero, meaning that each time step instantly converges once mesh changes have been handled.

**Reply by GiudGiud** [reply]
at 76539, is this when one of the blocks disappear by any chance?

**Reply by Edward-Eth** [reply]
It's not no, even if a block disappeared in the first simulation at 765390 seconds, the second simulation failed at a time value half of that as I halved my dt and it failed at the same step count. A run with memory post processors (total, max proc and per proc average) is still running but at 15k steps the memory uses are actually a hair under the memory use at t=0.

**Reply by GiudGiud** [reply]
Ok this rules out two potential causes. 
At this point I would remove components from the simulation one by one and try to find what causes a failure at 76539 time steps.

Is the mesh changing every time step?

**Reply by Edward-Eth** [reply]
There are no physics components to the simulation currently, just some variables with no assigned kernels. The mesh is changing most if not all time steps, with the current mesh+timings it's hard to be sure. I'm working out a more generic example I can use to test this out in a more controlled way.

**Reply by Edward-Eth** [reply]
So the memory usage postprocessor job just failed at the same point as the others. Here's the first and last memory usage postprocessor lines:
```
+----------------+----------------+----------------+----------------+
| time           | max_proc       | per_proc       | total          |
+----------------+----------------+----------------+----------------+
:                :                :                :                :
|   0.000000e+00 |   1.244000e+03 |   1.123990e+03 |   5.395130e+05 |
|   3.826900e+05 |   1.231000e+03 |   1.107627e+03 |   5.316610e+05 |
+----------------+----------------+----------------+----------------+
```
A normal time step looks like this:
```
Time Step 1, time = 5, dt = 5


Performing automatic scaling calculation


 0 Nonlinear |R| = 0.000000e+00

    |residual|_2 of individual variables:
                     disp_x: 0
                     disp_y: 0
                     disp_z: 0
                     Temp:   0

 Solve Converged!
 ```
 Whereas the frozen step looks like this:
 ```
 Time Step 76539, time = 382695, dt = 5


  Computing User Objects
    Handling Mesh Changes.......
```
It will continue printing dots to the output for as long as the job is left running, so something is still "working" in there, it's not a total freeze up.

Still trying to figure out a share-able MWE, got a few things on my plate atm so might be a post christmas thing potentially.

**Reply by GiudGiud** [reply]
Ok this rules out memory for sure again.
sounds good. We ll wait on the MWE to look at this then

**Comment by friedmud** [comment]
Are you running in parallel?  It could be that we're exhausting an MPI resource.

**Reply by Edward-Eth** [reply]
I'm running in parallel and with distributed mesh. How would I monitor MPI resources?

**Reply by hugary1995** [reply]
I don't think we set up the element activation UO to work with distributed mesh.

**Reply by Edward-Eth** [reply]
It's definitely working differently in distributed vs serial, the time to handle mesh changes is vastly longer with replicated mesh and does not change with the number of cores, whereas it scales perfectly with core count on distributed, and produces the expected result in terms of mesh changes.
Example of changing core count and mesh size on replicated mesh:
<img width="610" alt="Screenshot 2022-12-12 at 13 46 10" src="https://user-images.githubusercontent.com/64844573/207060719-a97af428-67df-4290-8c54-64393be56f30.png">
Example with distributed mesh:
<img width="615" alt="Screenshot 2022-12-12 at 13 47 54" src="https://user-images.githubusercontent.com/64844573/207061399-25b24b7d-b62f-4974-8840-ce9013a9c9a7.png">
EDIT: Apologies didn't notice that this somehow ended up dual axis, the distributed mesh is plotted against the right axis, showing that it is lower at all local DOF counts and decreases in line with local DOFs/elements/nodes

These graphs show that using distributed mesh is definitely accelerating the element activation object, meaning it must be running in parallel when distributed mesh is being used.

**Comment by Edward-Eth** [comment]
Post christmas estimate was correct for the MWE, but I've now run it and it also fails at the same number of time steps. To recap this means a mesh generated using an inhouse generation tool, with element activation timing governed by a CSV read into an aux-variable, and a mesh generated using GeneratedMeshGenerator with the activation timing governed by element z position fail at the exact same time step. The meshes have different numbers of elements and nodes, and previous testing has shown the time step size to be an irrelevant factor. What this points to I have no idea.

The MWE makes a single element base, and a 1 million element "tower", with the base starting in block 1 and the tower in block 0. This mesh is generated in the first file, MakeMesh.i, using the --mesh-only argument to generate an exodus file:

<details>
  <summary>MakeMesh.i</summary>

   ```
  [Mesh]
  [Tower]
    type = GeneratedMeshGenerator
    dim = 3
    elem_type = HEX8
    nx = 1
    ny = 1
    nz = 1000000
    xmax = 1
    ymax = 1
    zmax = 1000000
  []
  [Foundation]
    type = GeneratedMeshGenerator
    dim = 3
    element_type = HEX8
    nx = 1
    ny = 1
    nz = 1
    xmax = 1
    ymax = 1
    zmin = -1
    zmax = 0
  []
  [Stitch]
    type = StitchedMeshGenerator
    inputs = 'Tower Foundation'
    stitch_boundaries_pairs = 'back front'
  []
  [FoundationInBlock1]
    type = ParsedSubdomainMeshGenerator
    combinatorial_geometry = z<0
    block_id = 1
    input = Stitch
  []
  [SurfaceOfBlock1]
    type = SideSetsAroundSubdomainGenerator
    block = 1
    input = FoundationInBlock1
    new_boundary = 'surface'
  []
[]
  ```
</details>

This generated exodus file is then loaded into ActivateMesh.i, in which the elements in the tower block are slowly added into the foundation block, and variables added to them. This is analogous to a piece of material being added to the model, from a non-physical "placeholder" element to a physical one with variables and physics. This is the input file:

<details>
    <summary>ActivateMesh.i</summary>

```
# Tell MOOSE to ignore the physics-less deactivated domain.
[Problem]
  kernel_coverage_check = false
  material_coverage_check = false
[]

# Load the prepped mesh from an exodus file.
[Mesh]
  [Distributed_File]
    type = FileMeshGenerator
    file = ${Mesh_File}
    allow_renumbering = false
  []
  allow_renumbering = false
  parallel_type = DISTRIBUTED
[]

[Variables]
  [disp_x]
    block = 1
    order = FIRST
    family = LAGRANGE
  []
  [disp_y]
    block = 1
    order = FIRST
    family = LAGRANGE
  []
  [disp_z]
    block = 1
    order = FIRST
    family = LAGRANGE
  []
  [Cure]
    block = 1
    order = FIRST
    family = LAGRANGE
  []
[]

[AuxVariables]
  [Activation_Time]
    order = CONSTANT
    family = MONOMIAL
  []
  [Time_Until_Activation]
    order = CONSTANT
    family = MONOMIAL
  []
[]

[ICs]
  [Set_Activation_Time]
    type = FunctionIC
    variable = 'Activation_Time'
    function = z
  []
[]

[AuxKernels]
  [Activation_Countdown]
    type = ParsedAux
    variable = Time_Until_Activation
    args = 'Activation_Time'
    use_xyzt = true
    function = 'Activation_Time - t'
    execute_on = 'TIMESTEP_BEGIN TIMESTEP_END'
    block = 0
  []
[]

[UserObjects]
  [Activation]
    type = ActivateElementsCoupled
    execute_on = TIMESTEP_BEGIN
    coupled_var = Time_Until_Activation
    activate_value = 0.0
    activate_type = below
    active_subdomain_id = 1
    expand_boundary_name = 'surface'
    force_postaux = true
  []
[]

[Preconditioning]
  [SMP]
    type = SMP
    full = true
  []
[]

[Executioner]
  type = Transient

  start_time = 0

  end_time = 1220000
  
  dt = 5
  
  solve_type = 'NEWTON'
  automatic_scaling = true
[]

[Outputs]
  [Performance_Graph]
    type = PerfGraphOutput
    level = 1
    heaviest_branch = true
    heaviest_sections = 10
  []
[]
```
</details>

The files are run one after the other using a slurm script, key information is as follows:
- 5 Nodes
- 24 Cores per node
- 120 Cores total
- Run with mpirun

**Reply by GiudGiud** [reply]
i ll be delayed a bit in looking at this. 
Hopefully the issue is still visible with a few elements by changing the mesh, so I can look on a single node at most

**Reply by Edward-Eth** [reply]
Yeah I would assume given my experience to date that it has had the same issue on 40, 20, 10 and 5 nodes, and with two totally unrelated meshes, that it should occur on one node with a smaller mesh of the same time, maybe 100,000 elements with 1 activating per step.

**Reply by GiudGiud** [reply]
we really need to have a MWE. 1M elements is not a MWE. 
I've run it in serial with 100 elements up to 800,000 time steps without issue. 
Same 112,000 time steps with 2 procs with no issue
Same 93k time steps with 2 procs with distributed mesh no issue

**Reply by Edward-Eth** [reply]
Okay, I will look into creating something smaller that shows the issue I'm having.

**Reply by GiudGiud** [reply, marked as answer]
any luck with that?
